src/server/game/AI/SmartScripts/SmartScript.cpp

```diff
@@ -1182,10 +1182,14 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
         }
         case SMART_ACTION_SUMMON_CREATURE:
         {
-            WorldObject* summoner = GetBaseObjectOrUnit(unit);
+            EnumFlag<SmartActionSummonCreatureFlags> flags(static_cast<SmartActionSummonCreatureFlags>(e.action.summonCreature.flags));
+            bool preferUnit = flags.HasFlag(SmartActionSummonCreatureFlags::PreferUnit);
+            WorldObject* summoner = preferUnit ? (unit ? unit : GetBaseObject()) : GetBaseObjectOrUnit(unit);
```

Comment:
I do not like it works in quite nondeterministic way. 

The purpose of the flag is to use action invoker as the summoner. So why fallback to BaseObject when invoker is null?

I’d rather keep it simple and obvious - instead of `prefer unit`, imho it is better to make the flag `use invoker as summoner`. And if there is no invoker, then simply do not summon unit.

I do not see any reason why fallback to BaseObject will be good. If someone needs to use invoker as summoner, then he has valid reason not to use BaseObject (for instance making it personal spawn)

Reply:
Cool, I agree.
I'm keeping the other side (without the flag) with the fallback though, due to compatibility.